### PR TITLE
DIF Presentation Exchange Implementation in ACA-Py

### PR DESCRIPTION
**Pull request by shaangill025**
- https://identity.foundation/presentation-exchange/
- https://w3c.github.io/vc-data-model/

### CODE REVIEW

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+            to apply filtering on
+    Return:
+        Sequence of applicable VCRecords
+
+    """
+    result = []
+    for tmp_cred in credentials:
+        if (
+            constraints.subject_issuer is not None
+            and constraints.subject_issuer == "required"
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:193]: The `is not None` is redundant I guess?

```suggestion
            constraints.subject_issuer == "required"
```

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+        ):
+            continue
+
+        applicable = False
+        predicate = False
+        for tmp_field in constraints._fields:
+            applicable = await filter_by_field(tmp_field, tmp_cred)
+            if tmp_field.predicate:
+                if tmp_field.predicate == "required":
+                    predicate = True
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:204]:
```suggestion
            if tmp_field.predicate == "required":
            	predicate = True
```

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+    Args:
+        field: Field contains filtering spec
+        credential: credential to apply filtering on
+    Return:
+        bool
+
+    """
+    for tmp_path in field.paths:
+        tmp_jsonpath = parse(tmp_path)
+        match = tmp_jsonpath.find(json.loads(credential.value))
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:233]: You're parsing the stringified credential for every iteration of the loop. Also from what I can see, `filter_by_field` is called multiple times for every credential.

Preferably the already parsed credential is passed. But we can at least call `json.loads` outside of the `fields.path` loop

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+            for tmp_cred in tmp_creds_list:
+                if await trim_tmp_id(tmp_cred.given_id) not in tmp_dict:
+                    tmp_dict[await trim_tmp_id(tmp_cred.given_id)] = {}
+                tmp_dict[await trim_tmp_id(tmp_cred.given_id)][
+                    tmp_desc_id
+                ] = tmp_cred.given_id
+
+        if len(tmp_result.keys()) != 0:
+            nested_result.append(tmp_result)
+
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:656]: I'm sorry but I have no idea what's happening here. Could you maybe add some comments / more descriptive names or restructure a bit?

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+    exclude = {}
+    for k in tmp_dict.keys():
+        if not await is_len_applicable(req, len(tmp_dict[k])):
+            for desc_id in tmp_dict[k]:
+                exclude[desc_id + (tmp_dict[k][desc_id])] = {}
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:661]: Same here. It's really hard to follow

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
Thanks @shaangill025. The overall structure looks good to me, but the code lacks some comments / descriptive variable names and sometimes has some very complex code structure that make it a bit hard to follow what's happening.

So I've left some comments across the code, but didn't check the complete flow yet.

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch.py [hunk 6]
```diff
@@ -0,0 +1,1019 @@
...
+    )
+    descriptor_maps = fields.List(
+        fields.Nested(InputDescriptorMappingSchema),
+        required=False,
+        data_key="descriptor_map",
+    )
+
+
+# Union of str or dict
+class CustomValueField(fields.Field):
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch.py:921]: Oh and could you change the `PresentationExchError` to a `ValidationError` and move this to a more generic place? I need the same for some of the issue cred fields so would be nice to reuse this.

**Comment by shaangill025** [comment]
@TimoGlastra I have made the requested changes and added comments [especially in the apply_requirement function] to provide some clarity.

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch.py [hunk 7]
```diff
@@ -0,0 +1,1000 @@
...
+    @post_dump
+    def reformat_data(self, data, **kwargs):
+        """Support serialization of statuses according to DIF spec."""
+        if "status_active" in data:
+            tmp_dict = {}
+            tmp_dict["directive"] = data.get("status_active")
+            tmp_dict2 = data.get("statuses") or {}
+            tmp_dict2["active"] = tmp_dict
+            data["statuses"] = tmp_dict2
+            del data["status_active"]
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch.py:587]:
```suggestion
        if "status_active" in data:
            statuses = data.get("statuses", {})
            statuses["active"] = {
            	"directive": data.get("status_active")
            }
            data["statuses"] = statuses
            del data["status_active"]
```

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 8]
```diff
@@ -0,0 +1,828 @@
...
+        nested_req=[],
+    )
+    for submission_requirement in srs:
+        try:
+            requirement.nested_req.append(
+                await to_requirement(submission_requirement, descriptors)
+            )
+        except Exception as err:
+            raise PresentationExchError(
+                "Error creating requirement " f"inside to_requirement function, {err}"
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:140]:
```suggestion
                f"Error creating requirement inside to_requirement function, {err}"
```

**Comment by shaangill025** [comment]
Creating new pull request against animo's cwu-bbs-main branch.

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+            result[descriptor._id] = filtered
+
+    if len(descriptor_list) != 0:
+        # Applies min, max or count attributes of submission_requirement
+        if is_len_applicable(req, len(result)):
+            return result
+        return {}
+
+    nested_result = []
+    tmp_dict = {}
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:656]: The last tmp_dict. Maybe something like this?
```suggestion
    given_id_descriptors = {}
```

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+            for credential in credential_list:
+                if credential.given_id not in tmp_dict:
+                    tmp_dict[credential.given_id] = {}
+                tmp_dict[credential.given_id][descriptor_id] = {}
+
+        if len(result.keys()) != 0:
+            nested_result.append(result)
+
+    exclude = {}
+    for k in tmp_dict.keys():
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:681]: Then here:
```suggestion
    for given_id in given_id_descriptors.keys():
```

File: aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py [hunk 11]
```diff
@@ -0,0 +1,828 @@
...
+        # Check if number of applicable credentials
+        # does not meet requirement specification
+        if not is_len_applicable(req, len(tmp_dict[k])):
+            for descriptor_id in tmp_dict[k]:
+                # Add to exclude dict
+                # with cred.given_id + descriptor_id as key
+                exclude[descriptor_id + k] = {}
```
Comment (TimoGlastra) [inline aries_cloudagent/protocols/present_proof/presentation_exchange/pres_exch_handler.py:688]: And here:

```suggestion
            for descriptor_id in given_id_descriptors[given_id]:
                # Add to exclude dict
                # with descriptor_id + cred.given_id as key
                exclude[descriptor_id + given_id] = {}
```

**Review by TimoGlastra — APPROVED** [review]
This is perfect! Thanks @shaangill025 

I just added a comment for the last tmp_dict
